[telemetry] optimize tagger and dsd telemetry

vickenty opened the pull request:
### What does this PR do?

Avoid re-hashing cardinality tag over and over (relatively slow), by exposing and caching basic counters for each cardinality tag value (relatively fast).

### Motivation

Tag hashing inside `Counter.Inc()` implementation is visible in the profiles and takes a noticeable amount of time.

### Describe how to test your changes

Run agent with telemetry enabled and configure telemetry check in `conf.d/agent_telemetry.d/conf.yaml`:

```
init_config:
instances:
  - prometheus_url: http://localhost:5000/telemetry
    namespace: "datadog.agent"
    metrics:
      - "*"
```

Check that these metrics are populated:
- `datadog.agent.tagger_queries`
- `datadog.agent.tagger_fetches`
- `datadog.agent.dogstatsd.processed`

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

vickenty committed:
Factor out repeated code

vickenty committed:
Add a simple counter telemetry type

Simple counter does not support dynamic tags, and can be much faster,
which is a benefit on hot code paths.

ogaca-dd commented on pkg/tagger/telemetry/telemetry.go:
```diff
@@ -67,3 +72,21 @@ var (
+	if card <= collectors.HighCardinality && queriesForCardinality[card] != nil {
+		queriesForCardinality[card].Inc()
+	} else {
+		queries.Inc(collectors.TagCardinalityToString(card))
```
Can `card <= collectors.HighCardinality && queriesForCardinality[card] != nil` be false?

vickenty replied:
Yes, in case we introduce new cardinality level but do not update the initialization code above.

ogaca-dd commented on pkg/telemetry/counter.go:
```diff
@@ -35,6 +33,10 @@ type Counter interface {
+	// WithValues returns SimpleCounter for this metric with the given tag values.
+	WithValues(tagsValue ...string) SimpleCounter
+	// WithTags returns SimpleCounter for this metric with the given tqg values.
+	WithTags(tags map[string]string) SimpleCounter
```
`WithTags` is not used.

vickenty replied:
I added this for symmetry with prometheus code, and I plan to use it in other parts of the code.

ogaca-dd commented on pkg/telemetry/simple_counter.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// NewSimpleCounter creates a new SimpleCounter with default options.
+func NewSimpleCounter(subsystem, name, help string) SimpleCounter {
```
It seems to not be used.

ogaca-dd replied:
My understanding was that the interface `SimpleCounter` was used to not expose the return type of [WithValues](https://github.com/DataDog/datadog-agent/pull/8756/files/865c02eba9aeaad01a8f3ca9c733e7855fbe0680#diff-27f89cdbfd2dea42cc343309bd42a99714255410d48269b1cc03b9cd351b244fR37). In this case having a constructor is not common.

vickenty replied:
It is not only to hide the return type, `SimpleCounter` can be used on its own.

ogaca-dd commented on pkg/telemetry/simple_counter.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// NewSimpleCounterWithOpts creates a new SimpleCounter.
+func NewSimpleCounterWithOpts(subsystem, name, help string, opts Options) SimpleCounter {
```
It seems to not be used.

ogaca-dd commented on pkg/telemetry/prom_counter.go:
```diff
@@ -59,3 +59,13 @@ func (c *promCounter) Delete(tagsValue ...string) {
+}
+
+// Withtags returns SimpleCounter for this metric with the given tag values.
+func (c *promCounter) WithTags(tags map[string]string) SimpleCounter {
```
It seems to not be used.

vickenty committed:
[tagger] optimize fetches telemetry

Cache telemetry counters for each fetcher to avoid expensive hashmap
lookups and locks.

vickenty committed:
[tagger] optimize queries telemetry

vickenty committed:
[dsd] optimize server telemetry

ogaca-dd commented on pkg/tagger/telemetry/telemetry.go:
```diff
@@ -78,3 +81,54 @@ var (
+// NewFetcherTelemetry returns new instance of counters for the given fetcher name.
+func NewFetcherTelemetry(name string) FetcherTelemetry {
+	return FetcherTelemetry{
+		Success:  fetches.WithValues(name, FetchSuccess),
```
`FetchSuccess`, `FetchNotFound` and `FetchError` do not need to be exported.

ogaca-dd commented on pkg/tagger/telemetry/telemetry.go:
```diff
@@ -78,3 +81,54 @@ var (
+		EmptyEntityID: queries.WithValues(name, QueryEmptyEntityID),
+		EmptyTags:     queries.WithValues(name, QueryEmptyTags),
+		Success:       queries.WithValues(name, QuerySuccess),
+	}
```
`QueryEmptyEntityID`, `QueryEmptyTags` and `QuerySuccess` do not need to be exported.

ogaca-dd commented on pkg/telemetry/counter.go.orig:
```diff
@@ -0,0 +1,69 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
This file should not be committed.

vickenty committed:
remove file added by mistake

vickenty committed:
Unexport private constants

ogaca-dd reviewed (approved):
LGTM

remeh reviewed (approved):
Nice! Had you validated any improvements by looking at profiles by any chance?
Please make sure to have someone from container-integrations to double-check the `pkg/tagger` part but LGTM!

juliogreff reviewed (approved):
Tagger changes look good to me :shipit:
